storage/src/index/storage.rs

```diff
@@ -0,0 +1,180 @@
+            self.location = 0;
+            return Some(location);
+        }
+
+        match self.next {
+            Some(next) => {
+                self.location = next.location;
+                self.next = next.next.as_deref();
+                Some(next.location)
+            }
+            None => None,
+        }
+    }
+}
+
+impl Record {
+    fn iter(&self) -> LocationIterator {
+        LocationIterator {
+            location: self.location,
+            next: self.next.as_deref(),
+        }
+    }
+}
+
+pub struct Index<T: Translator, E: Metrics> {
+    translator: T,
+    map: HashMap<T::Key, Record>,
+    collisions: u64,
+    keys_pruned: Counter,
+
+    _metrics: std::marker::PhantomData<E>,
+}
+
+impl<T: Translator, E: Metrics> Index<T, E> {
+    /// Create a new index.
+    pub fn init(context: E, translator: T) -> Self {
+        let s = Self {
+            translator,
+            map: HashMap::new(),
+            collisions: 0,
+            keys_pruned: Counter::default(),
+            _metrics: std::marker::PhantomData,
+        };
+        context.register(
+            "keys_pruned",
+            "Number of keys pruned",
+            s.keys_pruned.clone(),
+        );
+
+        s
+    }
+
+    pub fn len(&self) -> usize {
+        self.map.len()
+    }
+
+    pub fn collisions(&self) -> u64 {
+        self.collisions
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.map.is_empty()
+    }
+
+    /// Insert a new record into the index.
+    pub fn insert(&mut self, key: &[u8], location: u64) {
+        let translated_key = self.translator.transform(key);
+
+        match self.map.entry(translated_key) {
+            Entry::Occupied(entry) => {
+                let entry: &mut Record = entry.into_mut();
+                entry.next = Some(Box::new(Record {
+                    location,
+                    next: entry.next.take(),
+                }));
+                self.collisions += 1;
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(Record {
+                    location,
+                    next: None,
+                });
+            }
+        };
+    }
+
+    pub fn get(&self, key: &[u8]) -> LocationIterator {
+        let translated_key = self.translator.transform(key);
+        let Some(cursor) = self.map.get(&translated_key) else {
+            return LocationIterator::empty();
+        };
+
+        cursor.iter()
+    }
+
+    /// Remove locations associated with the key that match the `prune` predicate.
+    pub fn remove(&mut self, key: &[u8], prune: impl Fn(u64) -> bool) {
+        let translated_key = self.translator.transform(key);
+        let head = match self.map.get_mut(&translated_key) {
+            Some(head) => head,
+            None => return,
+        };
+
+        // Advance the head of the linked list to the first entry that will be retained, if any.
+        loop {
+            if !prune(head.location) {
+                break;
```

Comment:
I think we want to continue here?

Reply:
continuing here would be an infinite loop... !prune means we want to retain the current record, so I think breaking is correct?